File: packages/permissions/src/core/engine.ts
/**
 * IAM Policy Engine - Core evaluation logic.
 * INTERNAL MODULE - use the `can` function from index.ts instead.
 */

import type {
  IEvaluationContext,
  IPolicyDocument,
  IStatement,
  TConditionBlock,
  TConditionOperator,
} from "../types";

/**
 * Condition operator functions
 */
const CONDITION_OPERATORS: Record<
  TConditionOperator,
  (actual: unknown, expected: unknown) => boolean
> = {
  StringEquals: (a, b) => String(a) === String(b),
  StringNotEquals: (a, b) => String(a) !== String(b),
  StringLike: (a, b) => matchWildcard(String(a), String(b)),
  StringNotLike: (a, b) => !matchWildcard(String(a), String(b)),
  NumericEquals: (a, b) => Number(a) === Number(b),
  NumericNotEquals: (a, b) => Number(a) !== Number(b),
  NumericLessThan: (a, b) => Number(a) < Number(b),
  NumericLessThanEquals: (a, b) => Number(a) <= Number(b),
  NumericGreaterThan: (a, b) => Number(a) > Number(b),
  NumericGreaterThanEquals: (a, b) => Number(a) >= Number(b),
  Bool: (a, b) => Boolean(a) === (String(b).toLowerCase() === "true"),
  Null: (a, b) => (a === null || a === undefined) === (String(b).toLowerCase() === "true"),
  "ForAnyValue:StringEquals": (a, b) => {
    const arr = Array.isArray(a) ? a : [a];
    return arr.some((v) => String(v) === String(b));
  },
  "ForAllValues:StringEquals": (a, b) => {
    const arr = Array.isArray(a) ? a : [a];
    return arr.every((v) => String(v) === String(b));
  },
};

/**
 * Match a string against a wildcard pattern
 * Supports * for any characters
 */
function matchWildcard(str: string, pattern: string): boolean {
  const regexPattern = pattern
    .replace(/[.+^${}()|[\]\\]/g, "\\$&") // Escape special regex chars
    .replace(/\*/g, ".*") // Convert * to .*
    .replace(/\?/g, "."); // Convert ? to .

  const regex = new RegExp(`^${regexPattern}$`, "i");
  return regex.test(str);
}

/**
 * Get a nested value from an object using dot notation
 * @example getNestedValue({a: {b: 1}}, "a.b") => 1
 */
function getNestedValue(obj: Record<string, unknown>, path: string): unknown {
  const parts = path.split(".");
  let current: unknown = obj;

  for (const part of parts) {
    if (current === null || current === undefined) {
      return undefined;
    }
    if (typeof current === "object" && part in current) {
      current = (current as Record<string, unknown>)[part];
    } else {
      return undefined;
    }
  }

  return current;
}

/**
 * Interpolate ${...} variables in a string
 * @example interpolateVariables("${actor.email}", context) => "user@example.com"
 */
function interpolateVariables(value: unknown, context: IEvaluationContext): unknown {
  if (typeof value !== "string") {
    return value;
  }

  return value.replace(/\$\{([^}]+)\}/g, (_, varPath: string) => {
    const varValue = getNestedValue(context as unknown as Record<string, unknown>, varPath);
    return varValue !== undefined ? String(varValue) : "";
  });
}

/**
 * Check if an action matches any of the patterns
 * Supports wildcards: "*" matches all, "project:*" matches all project actions
 */
function matchesAction(patterns: string[], action: string): boolean {
  for (const pattern of patterns) {
    if (pattern === "*") {
      return true;
    }
    if (matchWildcard(action, pattern)) {
      return true;
    }
  }
  return false;
}

/**
 * Evaluate conditions against context
 * All conditions must match (AND logic)
 */
function evaluateConditions(
  conditions: TConditionBlock | undefined,
  context: IEvaluationContext
): boolean {
  if (!conditions) {
    return true;
  }

  for (const [operator, conditionBlock] of Object.entries(conditions)) {
    const opFunc = CONDITION_OPERATORS[operator as TConditionOperator];
    if (!opFunc) {
      // Unknown operator - skip (fail open for forward compatibility)
      continue;
    }

    for (const [key, expectedValue] of Object.entries(conditionBlock || {})) {
      // Interpolate variables in expected value
      const interpolated = interpolateVariables(expectedValue, context);

      // Get actual value from context
      const actualValue = getNestedValue(context as unknown as Record<string, unknown>, key);

      try {
        if (!opFunc(actualValue, interpolated)) {
          return false;
        }
      } catch {
        // Type conversion failed - condition not met
        return false;
      }
    }
  }

  return true;
}

/**
 * Policy Engine class for evaluating permissions
 */
export class PolicyEngine {
  /**
   * Evaluate if an action is allowed based on policies and context.
   *
   * AWS-like evaluation logic:
   * 1. Default deny
   * 2. Collect all applicable statements
   * 3. If any explicit deny matches, return false
   * 4. If any allow matches (and no deny), return true
   * 5. Otherwise, return false (implicit deny)
   */
  evaluate(policies: IPolicyDocument[], action: string, context: IEvaluationContext): boolean {
    const allowStatements: IStatement[] = [];
    const denyStatements: IStatement[] = [];

    for (const policy of policies) {
      for (const statement of policy.statements) {
        if (matchesAction(statement.actions, action)) {
          if (evaluateConditions(statement.conditions, context)) {
            if (statement.effect === "allow") {
              allowStatements.push(statement);
            } else {
              denyStatements.push(statement);
            }
          }
        }
      }
    }

    // Explicit deny takes precedence
    if (denyStatements.length > 0) {
      return false;
    }

    // Check for allow
    if (allowStatements.length > 0) {
      return true;
    }

    // Implicit deny
    return false;
  }
}

// Singleton instance
const engineInstance = new PolicyEngine();

/**
 * Get the singleton PolicyEngine instance
 */
export function getEngine(): PolicyEngine {
  return engineInstance;
}

File: apps/api/plane/app/permissions/iam/engine.py
"""
IAM Policy Engine - Core evaluation logic.
INTERNAL MODULE - not exposed directly, use service.py functions.
"""
import fnmatch
import re
from typing import Any, Dict, List, Optional

from .types import Effect, EvaluationContext, PolicyDocument, Statement


class PolicyEngine:
    """
    Core policy evaluation engine.
    Evaluates policy documents against actions and contexts.

    This class is internal - use the `can()` function from service.py instead.
    """

    # Supported condition operators
    CONDITION_OPERATORS = {
        "StringEquals": lambda a, b: str(a) == str(b),
        "StringNotEquals": lambda a, b: str(a) != str(b),
        "StringLike": lambda a, b: fnmatch.fnmatch(str(a), str(b)),
        "StringNotLike": lambda a, b: not fnmatch.fnmatch(str(a), str(b)),
        "NumericEquals": lambda a, b: float(a) == float(b),
        "NumericNotEquals": lambda a, b: float(a) != float(b),
        "NumericLessThan": lambda a, b: float(a) < float(b),
        "NumericLessThanEquals": lambda a, b: float(a) <= float(b),
        "NumericGreaterThan": lambda a, b: float(a) > float(b),
        "NumericGreaterThanEquals": lambda a, b: float(a) >= float(b),
        "Bool": lambda a, b: bool(a) == (str(b).lower() == "true"),
        "Null": lambda a, b: (a is None) == (str(b).lower() == "true"),
        "ForAnyValue:StringEquals": lambda a, b: any(str(v) == str(b) for v in (a if isinstance(a, list) else [a])),
        "ForAllValues:StringEquals": lambda a, b: all(str(v) == str(b) for v in (a if isinstance(a, list) else [a])),
    }

    def evaluate(
        self,
        policies: List[PolicyDocument],
        action: str,
        context: EvaluationContext
    ) -> bool:
        """
        Evaluate if the action is allowed based on policies and context.

        AWS-like evaluation logic:
        1. Default deny
        2. Collect all applicable statements
        3. If any explicit deny matches, return False
        4. If any allow matches (and no deny), return True
        5. Otherwise, return False (implicit deny)

        Args:
            policies: List of policy documents to evaluate
            action: The action being performed (e.g., "project:delete")
            context: The evaluation context

        Returns:
            True if action is allowed, False otherwise
        """
        context_dict = context.to_dict()

        # Collect all matching statements
        allow_statements: List[Statement] = []
        deny_statements: List[Statement] = []

        for policy in policies:
            for statement in policy.statements:
                if self._matches_action(statement.actions, action):
                    if self._evaluate_conditions(statement.conditions, context_dict):
                        if statement.effect == Effect.ALLOW:
                            allow_statements.append(statement)
                        else:
                            deny_statements.append(statement)

        # Explicit deny takes precedence
        if deny_statements:
            return False

        # Check for allow
        if allow_statements:
            return True

        # Implicit deny
        return False

    def _matches_action(self, patterns: List[str], action: str) -> bool:
        """
        Check if action matches any of the patterns.
        Supports wildcards: "*" matches all, "project:*" matches all project actions.

        Args:
            patterns: List of action patterns (e.g., ["project:*", "issue:read"])
            action: The action to check (e.g., "project:delete")

        Returns:
            True if action matches any pattern
        """
        for pattern in patterns:
            if pattern == "*":
                return True
            if fnmatch.fnmatch(action, pattern):
                return True
        return False

    def _evaluate_conditions(
        self,
        conditions: Dict[str, Dict[str, Any]],
        context: Dict[str, Any]
    ) -> bool:
        """
        Evaluate all conditions against the context.
        All conditions must match (AND logic).

        Args:
            conditions: Condition block from statement
            context: Evaluation context as dictionary

        Returns:
            True if all conditions match
        """
        if not conditions:
            return True

        for operator, condition_block in conditions.items():
            if operator not in self.CONDITION_OPERATORS:
                # Unknown operator - skip (fail open for forward compatibility)
                continue

            op_func = self.CONDITION_OPERATORS[operator]

            for key, expected_value in condition_block.items():
                # Interpolate variables in expected value
                expected_value = self._interpolate_variables(expected_value, context)

                # Get actual value from context
                actual_value = self._get_nested_value(context, key)

                try:
                    if not op_func(actual_value, expected_value):
                        return False
                except (TypeError, ValueError):
                    # Type conversion failed - condition not met
                    return False

        return True

    def _interpolate_variables(self, value: Any, context: Dict[str, Any]) -> Any:
        """
        Interpolate ${...} variables in condition values.

        Example:
            "${actor.email}" -> "user@example.com"

        Args:
            value: The value to interpolate
            context: Context dictionary

        Returns:
            Interpolated value
        """
        if not isinstance(value, str):
            return value

        # Find all ${...} patterns
        pattern = r'\$\{([^}]+)\}'

        def replacer(match):
            var_path = match.group(1)
            var_value = self._get_nested_value(context, var_path)
            return str(var_value) if var_value is not None else ""

        return re.sub(pattern, replacer, value)

    def _get_nested_value(self, obj: Dict[str, Any], path: str) -> Any:
        """
        Get nested value from dictionary using dot notation.

        Example:
            _get_nested_value({"a": {"b": 1}}, "a.b") -> 1

        Args:
            obj: Dictionary to search
            path: Dot-separated path (e.g., "resource.group.name")

        Returns:
            Value at path or None if not found
        """
        parts = path.split(".")
        current = obj

        for part in parts:
            if isinstance(current, dict) and part in current:
                current = current[part]
            else:
                return None

        return current


# Singleton instance for reuse
_engine = PolicyEngine()


def get_engine() -> PolicyEngine:
    """Get the singleton PolicyEngine instance."""
    return _engine
